src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -539,59 +463,42 @@ RouterExecutorRun(QueryDesc *queryDesc, ScanDirection direction, long count)
 
 			if (isModificationQuery)
 			{
-				ExecuteSingleModifyTask(queryDesc, task, sendTuples);
+				bool sendTuples = multiPlan->hasReturning;
+				ExecuteSingleModifyTask(scanState, task, sendTuples);
 			}
 			else
 			{
-				ExecuteSingleSelectTask(queryDesc, task);
+				ExecuteSingleSelectTask(scanState, task);
 			}
 		}
 		else
 		{
-			ExecuteMultipleTasks(queryDesc, taskList, isModificationQuery,
+			bool sendTuples = multiPlan->hasReturning;
+			ExecuteMultipleTasks(scanState, taskList, isModificationQuery,
 								 sendTuples);
 		}
 
 		/* mark underlying query as having executed */
-		routerState->eof_underlying = true;
+		scanState->finishedUnderlyingScan = true;
 	}
 
 	/* if the underlying query produced output, return it */
-	if (routerState->tuplestorestate != NULL)
-	{
-		TupleDesc resultTupleDescriptor = queryDesc->tupDesc;
-		int64 returnedRows = 0;
-
-		/* return rows from the tuplestore */
-		returnedRows = ReturnRowsFromTuplestore(count, resultTupleDescriptor,
-												destination,
-												routerState->tuplestorestate);
 
-		/*
-		 * Count tuples processed, if this is a SELECT.  (For modifications
-		 * it'll already have been increased, as we want the number of
-		 * modified tuples, not the number of RETURNed tuples.)
-		 */
-		if (operation == CMD_SELECT)
-		{
-			estate->es_processed += returnedRows;
-		}
-	}
-
-out:
-
-	/* shutdown tuple receiver, if we started it */
-	if (sendTuples)
+	/*
+	 * FIXME: centralize this into function to be shared between router and
+	 * other executors?
+	 */
+	if (scanState->tuplestorestate != NULL)
```

Comment:
I think it'd make sense, because that'd allow us to support directional cursors more easily.